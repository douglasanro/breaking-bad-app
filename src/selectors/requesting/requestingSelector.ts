import { createSelector, ParametricSelector } from 'reselect';
import { IRequestingState } from 'stores/requesting/requestingModels';
import { rootState } from 'stores/rootStore';

export class SelectRequesting {
  public static requestingSelector(requestingState: IRequestingState, actionTypes: string[]): boolean {
    return actionTypes.some((actionType: string) => requestingState[actionType]);
  }
}

const requestingSelector: ParametricSelector<rootState, string[], boolean> = createSelector(
  (state: rootState) => state.requesting,
  (state: rootState, actionTypes: string[]) => actionTypes,
  SelectRequesting.requestingSelector
);

export default requestingSelector;
