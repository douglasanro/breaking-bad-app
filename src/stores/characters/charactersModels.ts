import ICharacter from 'models/ICharacter';

export type TCharacterFilterStatus = 'Alive' | 'Dead' | '';

export type TFilterTypes = 'status';

export interface ICharacterFilter {
  status: TCharacterFilterStatus;
};

export interface ICharacterState {
  list: ICharacter[];
  hasNextPage: boolean;
  filter: ICharacterFilter;
  searchTerm: string;
};

export interface IFetchCharacters {
  limit: number;
  offset?: number;
};
