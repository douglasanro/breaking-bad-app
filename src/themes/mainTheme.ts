import { unstable_createMuiStrictModeTheme as createMuiTheme } from '@material-ui/core';

const mainTheme = createMuiTheme({
  palette: {
    primary: {
      light: '#69A704',
      main: '#3A7116',
      dark: '#064500',
      contrastText: '#FFF',
    },
    secondary: {
      light: '#FFFF51',
      main: '#DFD400',
      dark: '#A9A300',
      contrastText: '#000',
    },
  },
  overrides: {
    MuiCssBaseline: {
      "@global": {
        "*, *::before, *::after": {
          boxSizing: "content-box",
        },
        body: {
          background: '#FFF'
        },
      },
    },
  },
});

export default mainTheme;
