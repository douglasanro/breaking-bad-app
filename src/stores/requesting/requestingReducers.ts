import IAction from 'models/IAction';
import { IRequestingState } from 'stores/requesting/requestingModels';

export const requestingInitialState = {}

const requestingReducers = (state = requestingInitialState, action: IAction<any>): IRequestingState => {
  const isFetchType = action.type.includes('FETCH_');

  if (!isFetchType) {
    return state;
  }

  const fetchName = action.type.replace(/_SUCCESS|_FAILURE/g, '_REQUEST');
  const isFinishedRequestType = ['_SUCCESS', '_FAILURE'].some(type => action.type.includes(type));

  return {
    ...state,
    [fetchName]: !isFinishedRequestType,
  };
};

export default requestingReducers;
