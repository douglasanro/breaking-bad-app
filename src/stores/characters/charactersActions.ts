import { action, createAsyncAction } from 'typesafe-actions';
import { ICharacterFilter } from './charactersModels';
import ICharacter from 'models/ICharacter';

export enum ECharactersActions {
  FETCH_CHARACTERS_REQUEST = '@characters/FETCH_CHARACTERS_REQUEST',
  FETCH_CHARACTERS_SUCCESS = '@characters/FETCH_CHARACTERS_SUCCESS',
  FETCH_CHARACTERS_FAILURE = '@characters/FETCH_CHARACTERS_FAILURE',
  CLEAR_CHARACTERS = '@characters/CLEAR_CHARACTERS',
  HAS_NEXT_PAGE = '@characters/HAS_NEXT_PAGE',
  SET_SEARCH_TERM = '@characters/SET_SEARCH_TERM',
  SET_FILTER = '@characters/SET_FILTER'
}

export const setCharacters = createAsyncAction(
  ECharactersActions.FETCH_CHARACTERS_REQUEST,
  ECharactersActions.FETCH_CHARACTERS_SUCCESS,
  ECharactersActions.FETCH_CHARACTERS_FAILURE,
)<undefined, ICharacter[], Error>();

export const clearCharacters = () => action(ECharactersActions.CLEAR_CHARACTERS);

export const setHasNextPage = (flag: boolean) => action(ECharactersActions.HAS_NEXT_PAGE, flag);

export const setSearchTerm = (term: string) => action(ECharactersActions.SET_SEARCH_TERM, term);

export const setFilter = ({ status }: ICharacterFilter) => action(ECharactersActions.SET_FILTER, { status });
