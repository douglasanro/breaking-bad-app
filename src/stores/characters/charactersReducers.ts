import { createReducer } from 'typesafe-actions';
import { ECharactersActions } from './charactersActions';
import { ICharacterState } from './charactersModels'

export const charactersInitialState: ICharacterState = {
  list: [],
  hasNextPage: true,
  filter: {
    status: ''
  },
  searchTerm: ''
};

const charactersReducers = createReducer(charactersInitialState, {
  [ECharactersActions.FETCH_CHARACTERS_SUCCESS]: (state, action) => ({
    ...state,
    list: [...new Set([...state.list, ...action.payload])]
  }),
  [ECharactersActions.CLEAR_CHARACTERS]: (state) => ({
    ...state,
    list: []
  }),
  [ECharactersActions.HAS_NEXT_PAGE]: (state, action) => ({
    ...state,
    hasNextPage: action.payload
  }),
  [ECharactersActions.SET_SEARCH_TERM]: (state, action) => ({
    ...state,
    searchTerm: action.payload
  }),
  [ECharactersActions.SET_FILTER]: (state, action) => ({
    ...state,
    filter: {...state.filter, ...action.payload}
  }),
});

export default charactersReducers;
