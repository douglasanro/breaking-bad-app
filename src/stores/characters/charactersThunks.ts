import { Dispatch } from 'redux';
import axios from 'axios';
import { setCharacters, setHasNextPage } from './charactersActions';
import { IFetchCharacters } from './charactersModels';
import ICharacter from 'models/ICharacter';
import { rootState } from 'stores/rootStore';
import { BREAKING_BAD_API_URL ,SERVICE } from 'environments';

const defaultOffset = 0;

export const getCharacters = ({ limit, offset = defaultOffset }: IFetchCharacters) => async ( dispatch: Dispatch, getState: () => rootState) => {
  const { characters } = getState();
  const { searchTerm } = characters;
  try {
    dispatch(setCharacters.request());
    const searchByName = searchTerm ? `&name=${searchTerm}` : '';
    const { data } = await axios.get(`${BREAKING_BAD_API_URL}${SERVICE.CHARACTERS()}?limit=${limit}&offset=${offset}${searchByName}`);

    if(!data.length) {
      dispatch(setHasNextPage(false));
      dispatch(setCharacters.success([]));
      return;
    }

    const effectData = data.map(({ status, ...otherArgs}: ICharacter) => {
      const currentStatus = ['Alive', 'Dead'][Number(status !== 'Alive')];

      return {...otherArgs, status: currentStatus };
    })

    dispatch(setHasNextPage(true));
    dispatch(setCharacters.success(effectData));
  } catch (error) {
    dispatch(setCharacters.failure(error));
  }
};
