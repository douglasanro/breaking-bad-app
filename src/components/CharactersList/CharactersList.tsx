import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Waypoint } from 'react-waypoint';
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import CharacterCard from 'components/CharacterCard';
import { getCharacters } from 'stores/characters/charactersThunks';
import { ECharactersActions, clearCharacters } from 'stores/characters/charactersActions';
import { rootState } from 'stores/rootStore';
import requestingSelector from 'selectors/requesting/requestingSelector';
import ICharacter from 'models/ICharacter';
import Typography from '@material-ui/core/Typography';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    container: {
      marginTop: theme.spacing(0),
      marginBottom: theme.spacing(0),
    },
    notFound: {
      marginTop: theme.spacing(4),
      textAlign: 'center',
    },
  })
);

const CharactersList: React.FC = () => {
  const classes = useStyles();
  const dispatch = useDispatch();

  const isRequesting = useSelector((state: rootState) =>
    requestingSelector(state, [ECharactersActions.FETCH_CHARACTERS_REQUEST])
  );
  const { list, hasNextPage, filter, searchTerm } = useSelector(({ characters }: rootState) => characters);

  const filteredList = list.filter((character) => {
    const hasFilter = Object.entries(filter).map(([key, value]) => {
      if (!value) {
        return true;
      }

      return character[key as keyof ICharacter] === value;
    });
    return hasFilter.every((filter) => filter);
  });

  const getMoreCharacters = () => {
    dispatch(getCharacters({ limit: 10, offset: list.length }));
  };

  useEffect(() => {
    if (!searchTerm) {
      dispatch(clearCharacters());
      dispatch(getCharacters({ limit: 10 }));
    }
  }, [dispatch, searchTerm]);

  if (!filteredList.length && !isRequesting) {
    return (
      <Typography component="p" variant="h4" className={classes.notFound}>
        No characters found. {searchTerm && 'Say my name correctly.'}
      </Typography>
    );
  }

  return (
    <Grid container spacing={4} className={classes.container}>
      {filteredList.map(({ char_id, name, birthday, img }, index) => (
        <Grid key={char_id} item xs={12} sm={6} md={4}>
          {hasNextPage && !isRequesting && index === filteredList.length - 1 && (
            <Waypoint onEnter={getMoreCharacters} />
          )}
          <CharacterCard name={name} birthday={birthday} img={img} />
        </Grid>
      ))}
    </Grid>
  );
};

export default CharactersList;
