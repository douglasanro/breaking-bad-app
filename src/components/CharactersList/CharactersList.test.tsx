import { render, cleanup } from '@testing-library/react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { ThemeProvider } from '@material-ui/core/styles';
import CharactersList from './CharactersList';
import { charactersInitialState } from 'stores/characters/charactersReducers';
import { requestingInitialState } from 'stores/requesting/requestingReducers';
import { mockCharacters } from 'setupTests';
import mainTheme from 'themes/mainTheme';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);
let store: ReturnType<typeof mockStore>;

const setupTest = () =>
  render(<CharactersList />, {
    wrapper: ({ children }) => (
      <Provider store={store}>
        <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
      </Provider>
    ),
  });

beforeEach(() => {
  store = mockStore({
    characters: {
      ...charactersInitialState,
      list: mockCharacters,
    },
    requesting: requestingInitialState,
  });
});

afterEach(() => {
  cleanup();
});

test('Render CharactersList correctly', () => {
  const { container } = setupTest();

  expect(container).toMatchSnapshot();
});

test('Render filtered CharactersList correctly', () => {
  store = mockStore({
    characters: {
      ...charactersInitialState,
      list: mockCharacters,
      filter: {
        status: 'Alive',
      },
    },
    requesting: requestingInitialState,
  });
  const { container } = setupTest();

  expect(container).toMatchSnapshot();
});

test('Render searched CharactersList correctly', () => {
  store = mockStore({
    characters: {
      ...charactersInitialState,
      list: mockCharacters,
      searchTerm: 'Walter',
    },
    requesting: requestingInitialState,
  });
  const { container } = setupTest();

  expect(container).toMatchSnapshot();
});

test('Render not found search correctly', () => {
  store = mockStore({
    characters: {
      ...charactersInitialState,
      searchTerm: 'Walter',
    },
    requesting: requestingInitialState,
  });
  const { container } = setupTest();

  expect(container).toMatchSnapshot();
});
