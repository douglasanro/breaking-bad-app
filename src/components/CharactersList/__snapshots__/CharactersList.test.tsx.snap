// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render CharactersList correctly 1`] = `
<div>
  <div
    class="MuiGrid-root makeStyles-container-1 MuiGrid-container MuiGrid-spacing-xs-4"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-card-3 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <img
          alt="Walter White"
          class="MuiCardMedia-root makeStyles-media-4 MuiCardMedia-media MuiCardMedia-img"
          src="https://images.amcnetworks.com/amc.com/wp-content/uploads/2015/04/cast_bb_700x1000_walter-white-lg.jpg"
          title="Walter White"
        />
        <div
          class="MuiCardContent-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
          >
            Walter White
          </h2>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
          >
            Age: 
            63 years
          </p>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
    >
      <span
        style="font-size: 0px;"
      />
      <div
        class="MuiPaper-root MuiCard-root makeStyles-card-3 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <img
          alt="Jesse Pinkman"
          class="MuiCardMedia-root makeStyles-media-4 MuiCardMedia-media MuiCardMedia-img"
          src="https://vignette.wikia.nocookie.net/breakingbad/images/9/95/JesseS5.jpg/revision/latest?cb=20120620012441"
          title="Jesse Pinkman"
        />
        <div
          class="MuiCardContent-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
          >
            Jesse Pinkman
          </h2>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
          >
            Age: 
            37 years
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Render filtered CharactersList correctly 1`] = `
<div>
  <div
    class="MuiGrid-root makeStyles-container-5 MuiGrid-container MuiGrid-spacing-xs-4"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
    >
      <span
        style="font-size: 0px;"
      />
      <div
        class="MuiPaper-root MuiCard-root makeStyles-card-7 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <img
          alt="Jesse Pinkman"
          class="MuiCardMedia-root makeStyles-media-8 MuiCardMedia-media MuiCardMedia-img"
          src="https://vignette.wikia.nocookie.net/breakingbad/images/9/95/JesseS5.jpg/revision/latest?cb=20120620012441"
          title="Jesse Pinkman"
        />
        <div
          class="MuiCardContent-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
          >
            Jesse Pinkman
          </h2>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
          >
            Age: 
            37 years
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Render not found search correctly 1`] = `
<div>
  <p
    class="MuiTypography-root makeStyles-notFound-14 MuiTypography-h4"
  >
    No characters found. 
    Say my name correctly.
  </p>
</div>
`;

exports[`Render searched CharactersList correctly 1`] = `
<div>
  <div
    class="MuiGrid-root makeStyles-container-9 MuiGrid-container MuiGrid-spacing-xs-4"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-card-11 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <img
          alt="Walter White"
          class="MuiCardMedia-root makeStyles-media-12 MuiCardMedia-media MuiCardMedia-img"
          src="https://images.amcnetworks.com/amc.com/wp-content/uploads/2015/04/cast_bb_700x1000_walter-white-lg.jpg"
          title="Walter White"
        />
        <div
          class="MuiCardContent-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
          >
            Walter White
          </h2>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
          >
            Age: 
            63 years
          </p>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
    >
      <span
        style="font-size: 0px;"
      />
      <div
        class="MuiPaper-root MuiCard-root makeStyles-card-11 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <img
          alt="Jesse Pinkman"
          class="MuiCardMedia-root makeStyles-media-12 MuiCardMedia-media MuiCardMedia-img"
          src="https://vignette.wikia.nocookie.net/breakingbad/images/9/95/JesseS5.jpg/revision/latest?cb=20120620012441"
          title="Jesse Pinkman"
        />
        <div
          class="MuiCardContent-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
          >
            Jesse Pinkman
          </h2>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
          >
            Age: 
            37 years
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
